src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -52,67 +76,176 @@ public void AddDependencyLocation(string fullPath)
+                // lived processes like VS and VBCSCompiler will see the same location added as
+                // a dependency many times. Each time have to assume there is new content on disk
+                // that needs to be considered.
+                _analyzerAssemblyInfoMap[fullPath] = null;
             }
         }
 
-        public Assembly LoadFromPath(string fullPath)
+        public Assembly LoadFromPath(string originalAnalyzerPath)
         {
-            CompilerPathUtilities.RequireAbsolutePath(fullPath, nameof(fullPath));
-            return LoadFromPathUnchecked(fullPath);
-        }
+            CompilerPathUtilities.RequireAbsolutePath(originalAnalyzerPath, nameof(originalAnalyzerPath));
 
-        #endregion
+            (AssemblyName? assemblyName, string _) = GetAssemblyInfoForPath(originalAnalyzerPath);
+
+            // Not a managed assembly, nothing else to do
+            if (assemblyName is null)
+            {
+                throw new ArgumentException("Not a valid assembly");
+            }
+
+            try
+            {
+                return Load(assemblyName, originalAnalyzerPath);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidOperationException($"Unable to load {assemblyName.Name}", ex);
+            }
+        }
 
         /// <summary>
-        /// Returns the cached assembly for fullPath if we've done a load for this path before, or calls <see cref="LoadFromPathUncheckedImpl"/> if
-        /// it needs to be loaded. This method skips the check in release builds that the path is an absolute path, hence the "Unchecked" in the name.
+        /// Get the <see cref="AssemblyName"/> and the path it should be loaded from for the given original 
+        /// analyzer path
         /// </summary>
-        protected Assembly LoadFromPathUnchecked(string fullPath)
+        /// <remarks>
+        /// This is used in the implementation of the loader instead of <see cref="AssemblyName.GetAssemblyName(string)"/>
+        /// because we only want information for registered paths. Using unregistered paths inside the
+        /// implementation should result in errors.
+        /// </remarks>
+        protected (AssemblyName? AssemblyName, string RealAssemblyPath) GetAssemblyInfoForPath(string originalAnalyzerPath)
         {
-            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
-
-            // Check if we have already loaded an assembly from the given path.
-            Assembly? loadedAssembly = null;
             lock (_guard)
             {
-                if (_loadedAssembliesByPath.TryGetValue(fullPath, out var existingAssembly))
+                if (!_analyzerAssemblyInfoMap.TryGetValue(originalAnalyzerPath, out var tuple))
                 {
-                    loadedAssembly = existingAssembly;
+                    throw new InvalidOperationException();
+                }
+
+                if (tuple is { } info)
```

Comment:
nit: we don't need a new local. Maybe `tuple` to `foundInfo` and then just do a `is not null` here.

Reply:
Correcth but then I have to do `tuple.Value.` in all the calls and I just prefer the local that doesn't have it.

Reply:
Thanks. I'd missed this was a nullable tuple.